File: packages/domain-tasks/src/parsing/quick-add.ts
/**
 * The quick-add grammar: `@date`, `every …`, `!priority`, `+project`, `#tag`,
 * `[[note]]` (spec 004 D1: English only, `@` prefix required for dates).
 *
 * Offsets are JavaScript string indices — UTF-16 code units — and the
 * conformance vectors pin them that way, so a port indexes the same units.
 * `now` is the caller's clock; nothing here reads one.
 */
import { parseNaturalDate } from './natural-date.ts'
import { findRepeatPhrase, firstOccurrenceFor } from './repeat-phrase.ts'
import type { Priority, QuickAddProject, RepeatConfig } from './types.ts'

// ============================================================================
// TYPES
// ============================================================================

export interface ParsedQuickAdd {
  title: string
  dueDate: Date | null
  /** "HH:MM", only when the natural-language date carried a time. */
  dueTime: string | null
  priority: Priority
  projectId: string | null
  repeat: RepeatConfig | null
  /** Every `#tag` in the input, in order, with the typed casing kept. */
  tags: string[]
  /**
   * The titles inside `[[…]]`, in order. The parser only sees text, so the
   * surface maps these to note ids (see CaptureBar).
   */
  noteTitles: string[]
}

/** A stretch of the input that carries syntax rather than title text. */
export type QuickAddSpanKind = 'priority' | 'project' | 'tag' | 'noteLink' | 'datePhrase' | 'repeat'

export interface QuickAddSpan {
  start: number
  /** Exclusive. */
  end: number
  kind: QuickAddSpanKind
}

// ============================================================================
// MARKER GRAMMAR
// ============================================================================

/**
 * `!` starts a word and needs at least one letter after it, so prose keeps its
 * punctuation: "Ship it!" and "Wow!!" are titles, not priorities.
 */
const PRIORITY_PATTERN = /(?:^|\s)(![a-zA-Z]+)/g

/** `+` starts a word, so "1+2" and "C++" never read as a project. */
const PROJECT_PATTERN = /(?:^|\s)(\+[\w-]+)/g

/**
 * The note editor's tag grammar verbatim, nesting included — see
 * `HASH_TAG_PATTERN` in `components/note/content-area/hash-tag.tsx`, which also
 * requires the `#` to start a word. `C#` and `issue#12` are therefore prose.
 */
const TAG_PATTERN = /(?:^|\s)(#[a-zA-Z0-9][a-zA-Z0-9_-]*(?:\/[a-zA-Z0-9][a-zA-Z0-9_-]*)*)/g

/** `[[Note title]]` — note titles are arbitrary, so anything but the brackets. */
const NOTE_LINK_PATTERN = /\[\[([^[\]\n]+)\]\]/g

interface MarkerRun {
  start: number
  /** Exclusive. */
  end: number
  /** The run including its sigil, e.g. `!high`. */
  text: string
  /** The run without its sigil, e.g. `high`. */
  value: string
}

/**
 * Every run of `pattern` in `input`. The patterns above match a leading space
 * to anchor the sigil to a word start, so the run itself is capture group 1.
 */
const findMarkerRuns = (input: string, pattern: RegExp): MarkerRun[] => {
  const runs: MarkerRun[] = []
  for (const match of input.matchAll(pattern)) {
    const text = match[1]
    const start = match.index + match[0].length - text.length
    runs.push({ start, end: start + text.length, text, value: text.slice(1) })
  }
  return runs
}

export interface NoteLinkMatch {
  start: number
  /** Exclusive. */
  end: number
  /** The text between the brackets. */
  title: string
}

/** Every finished `[[…]]` run. An unclosed `[[` is still being typed. */
export const findNoteLinks = (input: string): NoteLinkMatch[] => {
  const links: NoteLinkMatch[] = []
  for (const match of input.matchAll(NOTE_LINK_PATTERN)) {
    links.push({
      start: match.index,
      end: match.index + match[0].length,
      title: match[1].trim()
    })
  }
  return links
}

/**
 * A note title is user text and may well contain a sigil — `[[Q3 #launch]]`
 * links a note, it does not tag the task. Markers inside a link are ignored.
 */
const isInsideLink = (links: NoteLinkMatch[], index: number): boolean =>
  links.some((link) => index >= link.start && index < link.end)

// ============================================================================
// NATURAL-LANGUAGE DATE PHRASES (@tomorrow, @next wednesday, @dec 20 at 3pm)
// ============================================================================

export interface DatePhraseMatch {
  /** Index of the `@` in the scanned input. */
  start: number
  /** Exclusive end index of the phrase. */
  end: number
  /** The matched text including the `@` (what the pill paints over). */
  text: string
  date: Date
  /** "HH:MM" when the phrase carried a time, null otherwise. */
  time: string | null
}

/** `@` plus at most this many words — the cap on the greedy phrase scan. */
const MAX_DATE_PHRASE_WORDS = 4

/**
 * Find the first `@…` run that reads as a date, matching the note editor's
 * `@`-mention grammar: the `@` starts a word and the phrase can span several
 * words. Longest match wins, so "@next wednesday call bob" keeps "call bob" in
 * the title.
 */
export const findDatePhrase = (input: string, now: Date): DatePhraseMatch | null => {
  for (const at of input.matchAll(/@/g)) {
    const start = at.index
    // A mention starts a word — "a@b" is not one.
    if (start > 0 && !/\s/.test(input[start - 1])) continue

    const rest = input.slice(start + 1)
    if (!rest || /^\s/.test(rest)) continue

    const wordEnds = [...rest.matchAll(/\S+/g)]
      .slice(0, MAX_DATE_PHRASE_WORDS)
      .map((word) => word.index + word[0].length)

    for (let count = wordEnds.length; count >= 1; count--) {
      const phrase = rest.slice(0, wordEnds[count - 1])
      const parsed = parseNaturalDate(phrase, now)
      if (parsed.success) {
        return {
          start,
          end: start + 1 + phrase.length,
          text: `@${phrase}`,
          date: parsed.result.date,
          time: parsed.result.time
        }
      }
    }
  }
  return null
}

// ============================================================================
// PRIORITY PARSING
// ============================================================================

const priorityMap: Record<string, Priority> = {
  urgent: 'urgent',
  u: 'urgent',
  high: 'high',
  h: 'high',
  medium: 'medium',
  med: 'medium',
  m: 'medium',
  low: 'low',
  l: 'low',
  none: 'none',
  n: 'none'
}

/**
 * Parse priority keyword to Priority value
 */
export const parsePriorityKeyword = (keyword: string): Priority | null => {
  const lower = keyword.toLowerCase().trim()
  return priorityMap[lower] || null
}

// ============================================================================
// PROJECT PARSING
// ============================================================================

/**
 * Find project by name or ID (case-insensitive)
 */
export const findProjectByName = (
  name: string,
  projects: readonly QuickAddProject[]
): string | null => {
  const lower = name.toLowerCase().trim()

  // Try exact ID match first
  const byId = projects.find((p) => p.id.toLowerCase() === lower)
  if (byId) return byId.id

  // Try exact name match
  const byName = projects.find((p) => p.name.toLowerCase() === lower)
  if (byName) return byName.id

  // Try partial name match (starts with)
  const byPartial = projects.find((p) => p.name.toLowerCase().startsWith(lower))
  if (byPartial) return byPartial.id

  // Try kebab-case name match (e.g., "project-alpha" matches "Project Alpha")
  const kebabName = lower.replace(/-/g, ' ')
  const byKebab = projects.find((p) => p.name.toLowerCase() === kebabName)
  if (byKebab) return byKebab.id

  return null
}

// ============================================================================
// MAIN PARSER
// ============================================================================

/** Sorted, non-overlapping — an earlier span wins a contested stretch. */
const dropOverlaps = (spans: QuickAddSpan[]): QuickAddSpan[] => {
  const sorted = [...spans].sort((a, b) => a.start - b.start)
  const kept: QuickAddSpan[] = []
  for (const span of sorted) {
    if (kept.length > 0 && span.start < kept[kept.length - 1].end) continue
    kept.push(span)
  }
  return kept
}

const stripSpans = (input: string, spans: QuickAddSpan[]): string => {
  let title = ''
  let cursor = 0
  for (const span of spans) {
    title += input.slice(cursor, span.start)
    cursor = span.end
  }
  title += input.slice(cursor)
  return title.replace(/\s+/g, ' ').trim()
}

/**
 * Parse quick add input string with special syntax
 *
 * Syntax (the note editor's grammar, so both surfaces agree):
 * - Due date: @tomorrow, @next wednesday, @dec 20 at 3pm
 * - Repeat: every day, every weekday, every monday, every 2 weeks, every month
 * - Priority: !urgent, !high, !medium, !low
 * - Project: +project-name, +personal, +work
 * - Tag: #launch, #work/client — every tag in the input counts
 * - Note link: [[Roadmap]]
 *
 * Examples:
 * - "Buy groceries @today !high" → title: "Buy groceries", due: today, priority: high
 * - "Review PR +work @next friday" → title: "Review PR", project: work, due: next Friday
 * - "Water plants every 2 weeks" → title: "Water plants", repeats every second week
 */
export const parseQuickAdd = (
  input: string,
  projects: readonly QuickAddProject[],
  now: Date
): ParsedQuickAdd => {
  const spans: QuickAddSpan[] = []
  let dueDate: Date | null = null
  let dueTime: string | null = null
  let priority: Priority = 'none'
  let projectId: string | null = null

  // Note links first: they own their whole run, sigils inside included.
  const noteLinks = findNoteLinks(input)
  for (const link of noteLinks) {
    spans.push({ start: link.start, end: link.end, kind: 'noteLink' })
  }

  // Natural-language due date: @tomorrow, @next wednesday
  const datePhrase = findDatePhrase(input, now)
  if (datePhrase && !isInsideLink(noteLinks, datePhrase.start)) {
    dueDate = datePhrase.date
    dueTime = datePhrase.time
    spans.push({ start: datePhrase.start, end: datePhrase.end, kind: 'datePhrase' })
  }

  // Repeat: every monday, every 2 weeks. Anchored to the due date so a bare
  // "every month" repeats on the day the task is actually due.
  const repeatPhrase = findRepeatPhrase(input, dueDate ?? now, now)
  const repeat =
    repeatPhrase && !isInsideLink(noteLinks, repeatPhrase.start) ? repeatPhrase.config : null
  if (repeatPhrase && repeat) {
    spans.push({ start: repeatPhrase.start, end: repeatPhrase.end, kind: 'repeat' })
  }

  // Priority: !high. The first run that names a priority wins; "!nope" is prose.
  for (const run of findMarkerRuns(input, PRIORITY_PATTERN)) {
    if (isInsideLink(noteLinks, run.start)) continue
    const parsedPriority = parsePriorityKeyword(run.value)
    if (parsedPriority) {
      priority = parsedPriority
      spans.push({ start: run.start, end: run.end, kind: 'priority' })
      break
    }
  }

  // Project: +work. Unresolved "+foo" stays in the title.
  for (const run of findMarkerRuns(input, PROJECT_PATTERN)) {
    if (isInsideLink(noteLinks, run.start)) continue
    const foundProjectId = findProjectByName(run.value, projects)
    if (foundProjectId) {
      projectId = foundProjectId
      spans.push({ start: run.start, end: run.end, kind: 'project' })
      break
    }
  }

  // Tags: unlike the other markers, every one counts — a task takes several.
  const tags: string[] = []
  for (const run of findMarkerRuns(input, TAG_PATTERN)) {
    if (isInsideLink(noteLinks, run.start)) continue
    tags.push(run.value)
    spans.push({ start: run.start, end: run.end, kind: 'tag' })
  }

  // A repeat only rolls forward from a due date, so give an undated repeating
  // task its first occurrence.
  if (repeat && !dueDate) {
    dueDate = firstOccurrenceFor(repeat, now)
  }

  return {
    title: stripSpans(input, dropOverlaps(spans)),
    dueDate,
    dueTime,
    priority,
    projectId,
    repeat,
    tags,
    noteTitles: noteLinks.map((link) => link.title).filter(Boolean)
  }
}

/**
 * The stretches of `input` that carry syntax, for the token overlay. Same
 * source of truth as {@link parseQuickAdd}, so a highlighted phrase is always
 * one that actually left the title.
 */
export const findQuickAddSpans = (input: string, now: Date): QuickAddSpan[] => {
  const spans: QuickAddSpan[] = []

  const noteLinks = findNoteLinks(input)
  for (const link of noteLinks) {
    spans.push({ start: link.start, end: link.end, kind: 'noteLink' })
  }

  const datePhrase = findDatePhrase(input, now)
  if (datePhrase && !isInsideLink(noteLinks, datePhrase.start)) {
    spans.push({ start: datePhrase.start, end: datePhrase.end, kind: 'datePhrase' })
  }

  const repeatPhrase = findRepeatPhrase(input, now, now)
  if (repeatPhrase && !isInsideLink(noteLinks, repeatPhrase.start)) {
    spans.push({ start: repeatPhrase.start, end: repeatPhrase.end, kind: 'repeat' })
  }

  // The sigil forms are painted as soon as they are typed, parseable or not —
  // an unfinished "!hi" or an unknown "+foo" still reads as syntax.
  const sigilKinds: [RegExp, QuickAddSpanKind][] = [
    [PRIORITY_PATTERN, 'priority'],
    [PROJECT_PATTERN, 'project'],
    [TAG_PATTERN, 'tag']
  ]
  for (const [pattern, kind] of sigilKinds) {
    for (const run of findMarkerRuns(input, pattern)) {
      if (isInsideLink(noteLinks, run.start)) continue
      spans.push({ start: run.start, end: run.end, kind })
    }
  }

  return dropOverlaps(spans)
}

// ============================================================================
// PREVIEW HELPERS
// ============================================================================

/**
 * Check if input has any special syntax
 */
export const hasSpecialSyntax = (input: string, now: Date): boolean => {
  // Scanned with matchAll rather than `.test()`: these patterns are global and
  // module-level, so `.test()` would carry `lastIndex` between calls.
  return (
    findMarkerRuns(input, PRIORITY_PATTERN).length > 0 ||
    findMarkerRuns(input, PROJECT_PATTERN).length > 0 ||
    findMarkerRuns(input, TAG_PATTERN).length > 0 ||
    findNoteLinks(input).length > 0 ||
    findDatePhrase(input, now) !== null || // @tomorrow
    findRepeatPhrase(input, now, now) !== null // every monday
  )
}

// ============================================================================
// GHOST COMPLETION FOR REPEAT PHRASES
// ============================================================================

/**
 * The cadences the ghost completes to. Ordered so the shortest useful phrase
 * wins an ambiguous prefix ("every w" → weekday, the commonest routine).
 */
const REPEAT_PHRASES = [
  'every day',
  'every weekday',
  'every week',
  'every 2 weeks',
  'every weekend',
  'every month',
  'every year'
]

/**
 * Canonical cadence for a half-typed "every …", or null when nothing completes
 * it. Case-insensitive superstring of `query`, like the date completions.
 */
export const predictRepeatCompletion = (query: string): string | null => {
  const lower = query.toLowerCase()
  return REPEAT_PHRASES.find((phrase) => phrase.startsWith(lower)) ?? null
}

File: crates/memry-core/src/domain/task_parse/quick_add.rs
//! `parseQuickAdd`: the quick-add grammar, `@date`, `every ...`, `!priority`,
//! `+project`, `#tag`, `[[note]]` (spec 004 D1: English only, `@` prefix
//! required for dates), ported from
//! `packages/domain-tasks/src/parsing/quick-add.ts`, quirks included.
//!
//! Every offset handed out is a UTF-16 code unit index, end exclusive, as the
//! TypeScript and the vectors index strings. Internally offsets are UTF-8 byte
//! indices, converted at the API edge. `now` is the caller's local clock;
//! nothing here reads one.

use std::ops::Range;

use super::natural_date::{is_js_whitespace, js_trim, parse_natural_date};
use super::repeat_phrase::{
    collapse_whitespace, find_repeat_phrase_range, is_ascii_word, utf16_index, word_ends,
};
use crate::domain::calendar::{CivilDate, LocalDateTime};
use crate::domain::recurrence::first_occurrence_for;
use crate::domain::repeat_config::RepeatConfig;
use crate::domain::task_filter::Priority;

/// A project the `+name` marker can resolve to.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct QuickAddProject {
    pub id: String,
    pub name: String,
    /// Carried for the caller; desktop resolves archived projects too.
    pub is_archived: bool,
}

/// What a quick-add input captured.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedQuickAdd {
    pub title: String,
    pub due_date: Option<CivilDate>,
    /// `"HH:MM"`, only when the date phrase carried a time.
    pub due_time: Option<String>,
    pub priority: Priority,
    pub project_id: Option<String>,
    pub repeat: Option<RepeatConfig>,
    /// Every `#tag` in the input, in order, with the typed casing kept.
    pub tags: Vec<String>,
    /// The non-empty titles inside `[[...]]`, in order.
    pub note_titles: Vec<String>,
}

/// What a stretch of syntax is.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SpanKind {
    Priority,
    Project,
    Tag,
    NoteLink,
    DatePhrase,
    Repeat,
}

impl SpanKind {
    /// Desktop's `QuickAddSpanKind` name.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Priority => "priority",
            Self::Project => "project",
            Self::Tag => "tag",
            Self::NoteLink => "noteLink",
            Self::DatePhrase => "datePhrase",
            Self::Repeat => "repeat",
        }
    }
}

/// A stretch of the input that carries syntax rather than title text.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct QuickAddSpan {
    /// UTF-16 index.
    pub start: usize,
    /// Exclusive UTF-16 index.
    pub end: usize,
    pub kind: SpanKind,
}

/// A finished `[[...]]` run.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NoteLinkMatch {
    /// UTF-16 index of the first `[`.
    pub start: usize,
    /// Exclusive UTF-16 index after the last `]`.
    pub end: usize,
    /// The text between the brackets, trimmed.
    pub title: String,
}

/// An `@...` run that reads as a date.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DatePhraseMatch {
    /// UTF-16 index of the `@`.
    pub start: usize,
    /// Exclusive UTF-16 end index of the phrase.
    pub end: usize,
    /// The matched text including the `@`.
    pub text: String,
    pub date: CivilDate,
    /// `"HH:MM"` when the phrase carried a time.
    pub time: Option<String>,
}

/// `@` plus at most this many words: the cap on the greedy phrase scan.
const MAX_DATE_PHRASE_WORDS: usize = 4;

/// The cadences the ghost completes to, shortest useful phrase first.
const REPEAT_PHRASES: [&str; 7] = [
    "every day",
    "every weekday",
    "every week",
    "every 2 weeks",
    "every weekend",
    "every month",
    "every year",
];

/// A span in UTF-8 byte offsets.
#[derive(Debug, Clone)]
struct RawSpan {
    range: Range<usize>,
    kind: SpanKind,
}

/// A finished `[[...]]` run in UTF-8 byte offsets.
struct RawNoteLink {
    range: Range<usize>,
    title: String,
}

/// A date phrase in UTF-8 byte offsets.
struct RawDatePhrase {
    range: Range<usize>,
    date: CivilDate,
    time: Option<String>,
}

/// The sigil markers, each `/(?:^|\s)(<run>)/g`.
#[derive(Debug, Clone, Copy)]
enum Marker {
    /// `![a-zA-Z]+`: "Ship it!" and "Wow!!" stay prose.
    Priority,
    /// `\+[\w-]+`: "1+2" and "C++" never read as a project.
    Project,
    /// `#[a-zA-Z0-9][a-zA-Z0-9_-]*(?:\/[a-zA-Z0-9][a-zA-Z0-9_-]*)*`, the note
    /// editor's tag grammar: "C#" and "issue#12" are prose.
    Tag,
}

impl Marker {
    /// Byte length of the marker run at the start of `text`, sigil included.
    fn run_len(self, text: &str) -> Option<usize> {
        let bytes = text.as_bytes();
        let count_from = |from: usize, accept: fn(u8) -> bool| {
            bytes
                .get(from..)
                .map_or(0, |tail| tail.iter().take_while(|&&b| accept(b)).count())
        };
        let len = match self {
            Self::Priority if bytes.first() == Some(&b'!') => {
                1 + count_from(1, |b| b.is_ascii_alphabetic())
            }
            Self::Project if bytes.first() == Some(&b'+') => {
                1 + count_from(1, |b| is_ascii_word(b) || b == b'-')
            }
            Self::Tag if bytes.first() == Some(&b'#') => {
                let mut len = 1;
                loop {
                    if !bytes.get(len).is_some_and(u8::is_ascii_alphanumeric) {
                        break;
                    }
                    len += 1 + count_from(len + 1, |b| is_ascii_word(b) || b == b'-');
                    let nested = bytes.get(len) == Some(&b'/')
                        && bytes.get(len + 1).is_some_and(u8::is_ascii_alphanumeric);
                    if !nested {
                        break;
                    }
                    len += 1;
                }
                len
            }
            _ => return None,
        };
        (len > 1).then_some(len)
    }

    /// Every run in `input`, in `matchAll` order: a match consumes its
    /// leading whitespace, and the scan resumes where the run ends.
    fn runs(self, input: &str) -> Vec<Range<usize>> {
        let mut runs = Vec::new();
        let mut pos = 0;
        while let Some(rest) = input.get(pos..) {
            let Some(first) = rest.chars().next() else {
                break;
            };
            let run_start = if pos == 0 && self.run_len(rest).is_some() {
                Some(pos)
            } else if is_js_whitespace(first) {
                Some(pos + first.len_utf8())
            } else {
                None
            };
            let run = run_start.and_then(|start| {
                let len = self.run_len(input.get(start..)?)?;
                Some(start..start + len)
            });
            match run {
                Some(run) => {
                    pos = run.end;
                    runs.push(run);
                }
                None => pos += first.len_utf8(),
            }
        }
        runs
    }
}

/// Every finished `[[...]]` run (`/\[\[([^[\]\n]+)\]\]/g`). An unclosed `[[`
/// is still being typed.
pub fn find_note_links(input: &str) -> Vec<NoteLinkMatch> {
    raw_note_links(input)
        .into_iter()
        .map(|link| NoteLinkMatch {
            start: utf16_index(input, link.range.start),
            end: utf16_index(input, link.range.end),
            title: link.title,
        })
        .collect()
}

fn raw_note_links(input: &str) -> Vec<RawNoteLink> {
    let bytes = input.as_bytes();
    let mut links = Vec::new();
    let mut pos = 0;
    while pos + 1 < bytes.len() {
        if bytes.get(pos..pos + 2) != Some(b"[[".as_slice()) {
            pos += 1;
            continue;
        }
        let inner_start = pos + 2;
        let inner_len = bytes.get(inner_start..).map_or(0, |tail| {
            tail.iter()
                .take_while(|&&b| !matches!(b, b'[' | b']' | b'\n'))
                .count()
        });
        let inner_end = inner_start + inner_len;
        let closed = inner_len > 0 && bytes.get(inner_end..inner_end + 2) == Some(b"]]".as_slice());
        match input.get(inner_start..inner_end).filter(|_| closed) {
            Some(inner) => {
                links.push(RawNoteLink {
                    range: pos..inner_end + 2,
                    title: js_trim(inner).to_owned(),
                });
                pos = inner_end + 2;
            }
            None => pos += 1,
        }
    }
    links
}

/// A note title may well contain a sigil: `[[Q3 #launch]]` links a note, it
/// does not tag the task. Markers inside a link are ignored.
fn is_inside_link(links: &[RawNoteLink], index: usize) -> bool {
    links.iter().any(|link| link.range.contains(&index))
}

/// Finds the first `@...` run that reads as a date. The `@` starts a word and
/// the phrase spans up to four words, longest match first, so "@next
/// wednesday call bob" keeps "call bob" in the title.
pub fn find_date_phrase(input: &str, now: LocalDateTime) -> Option<DatePhraseMatch> {
    let phrase = raw_date_phrase(input, now)?;
    Some(DatePhraseMatch {
        start: utf16_index(input, phrase.range.start),
        end: utf16_index(input, phrase.range.end),
        text: input.get(phrase.range)?.to_owned(),
        date: phrase.date,
        time: phrase.time,
    })
}

fn raw_date_phrase(input: &str, now: LocalDateTime) -> Option<RawDatePhrase> {
    input.match_indices('@').find_map(|(start, _)| {
        // A mention starts a word: "a@b" is not one.
        let before = input.get(..start)?.chars().next_back();
        if before.is_some_and(|c| !is_js_whitespace(c)) {
            return None;
        }
        let rest = input.get(start + 1..)?;
        if rest.chars().next().is_none_or(is_js_whitespace) {
            return None;
        }
        word_ends(rest, MAX_DATE_PHRASE_WORDS)
            .into_iter()
            .rev()
            .find_map(|end| {
                let parsed = parse_natural_date(rest.get(..end)?, now).ok()?;
                Some(RawDatePhrase {
                    range: start..start + 1 + end,
                    date: parsed.date,
                    time: parsed.time,
                })
            })
    })
}

/// `parsePriorityKeyword`: `urgent`/`u`, `high`/`h`, `medium`/`med`/`m`,
/// `low`/`l`, `none`/`n`, any case, surrounding whitespace ignored.
///
/// An inherited `Object.prototype` key ("constructor") is truthy in the
/// TypeScript map lookup; here it is simply not a priority.
pub fn parse_priority_keyword(keyword: &str) -> Option<Priority> {
    let lower = keyword.to_lowercase();
    Some(match js_trim(&lower) {
        "urgent" | "u" => Priority::Urgent,
        "high" | "h" => Priority::High,
        "medium" | "med" | "m" => Priority::Medium,
        "low" | "l" => Priority::Low,
        "none" | "n" => Priority::None,
        _ => return None,
    })
}

/// `findProjectByName`, case-insensitive: an exact id, an exact name, a name
/// prefix, then a kebab-case name ("project-alpha" is "Project Alpha").
/// Archived projects resolve like any other, as on desktop.
pub fn find_project_by_name(name: &str, projects: &[QuickAddProject]) -> Option<String> {
    let lowered = name.to_lowercase();
    let lower = js_trim(&lowered);
    let kebab_name = lower.replace('-', " ");
    let lower_name = |project: &QuickAddProject| project.name.to_lowercase();
    projects
        .iter()
        .find(|p| p.id.to_lowercase() == lower)
        .or_else(|| projects.iter().find(|p| lower_name(p) == lower))
        .or_else(|| projects.iter().find(|p| lower_name(p).starts_with(lower)))
        .or_else(|| projects.iter().find(|p| lower_name(p) == kebab_name))
        .map(|project| project.id.clone())
}

/// Sorted by start (stably), non-overlapping: an earlier span wins.
fn drop_overlaps(mut spans: Vec<RawSpan>) -> Vec<RawSpan> {
    spans.sort_by_key(|span| span.range.start);
    let mut kept: Vec<RawSpan> = Vec::new();
    for span in spans {
        if kept
            .last()
            .is_some_and(|last| span.range.start < last.range.end)
        {
            continue;
        }
        kept.push(span);
    }
    kept
}

/// The input without `spans`, whitespace runs collapsed and trimmed.
fn strip_spans(input: &str, spans: &[RawSpan]) -> String {
    let mut title = String::with_capacity(input.len());
    let mut cursor = 0;
    for span in spans {
        title.push_str(input.get(cursor..span.range.start).unwrap_or_default());
        cursor = span.range.end;
    }
    title.push_str(input.get(cursor..).unwrap_or_default());
    js_trim(&collapse_whitespace(&title)).to_owned()
}

fn link_spans(links: &[RawNoteLink]) -> Vec<RawSpan> {
    links
        .iter()
        .map(|link| RawSpan {
            range: link.range.clone(),
            kind: SpanKind::NoteLink,
        })
        .collect()
}

/// Parses quick-add input: `@tomorrow`, `every monday`, `!high`, `+work`,
/// `#tag`, `[[Note]]`. What is not syntax is the title.
pub fn parse_quick_add(
    input: &str,
    projects: &[QuickAddProject],
    now: LocalDateTime,
) -> ParsedQuickAdd {
    // Note links first: they own their whole run, sigils inside included.
    let links = raw_note_links(input);
    let mut spans = link_spans(&links);
    let mut due_date = None;
    let mut due_time = None;
    let mut priority = Priority::None;
    let mut project_id = None;

    if let Some(phrase) =
        raw_date_phrase(input, now).filter(|p| !is_inside_link(&links, p.range.start))
    {
        due_date = Some(phrase.date);
        due_time = phrase.time;
        spans.push(RawSpan {
            range: phrase.range,
            kind: SpanKind::DatePhrase,
        });
    }

    // Anchored to the due date, so a bare "every month" repeats on the day the
    // task is actually due.
    let anchor = due_date.unwrap_or_else(|| now.date());
    let repeat = find_repeat_phrase_range(input, anchor, now)
        .filter(|(range, _)| !is_inside_link(&links, range.start))
        .map(|(range, config)| {
            spans.push(RawSpan {
                range,
                kind: SpanKind::Repeat,
            });
            config
        });

    // The first run that names a priority wins; "!nope" is prose.
    for run in Marker::Priority.runs(input) {
        if is_inside_link(&links, run.start) {
            continue;
        }
        if let Some(parsed) = input
            .get(run.start + 1..run.end)
            .and_then(parse_priority_keyword)
        {
            priority = parsed;
            spans.push(RawSpan {
                range: run,
                kind: SpanKind::Priority,
            });
            break;
        }
    }

    // An unresolved "+foo" stays in the title.
    for run in Marker::Project.runs(input) {
        if is_inside_link(&links, run.start) {
            continue;
        }
        let name = input.get(run.start + 1..run.end).unwrap_or_default();
        if let Some(found) = find_project_by_name(name, projects) {
            project_id = Some(found);
            spans.push(RawSpan {
                range: run,
                kind: SpanKind::Project,
            });
            break;
        }
    }

    // Unlike the other markers, every tag counts.
    let mut tags = Vec::new();
    for run in Marker::Tag.runs(input) {
        if is_inside_link(&links, run.start) {
            continue;
        }
        tags.push(
            input
                .get(run.start + 1..run.end)
                .unwrap_or_default()
                .to_owned(),
        );
        spans.push(RawSpan {
            range: run,
            kind: SpanKind::Tag,
        });
    }

    // A repeat only rolls forward from a due date.
    if let (Some(config), None) = (&repeat, due_date) {
        due_date = Some(first_occurrence_for(config, now));
    }

    ParsedQuickAdd {
        title: strip_spans(input, &drop_overlaps(spans)),
        due_date,
        due_time,
        priority,
        project_id,
        repeat,
        tags,
        note_titles: links
            .into_iter()
            .map(|link| link.title)
            .filter(|title| !title.is_empty())
            .collect(),
    }
}

/// The stretches of `input` that carry syntax, for the token overlay, sorted
/// and non-overlapping. The sigil forms count as soon as they are typed,
/// parseable or not: an unfinished "!hi" or an unknown "+foo" is syntax.
pub fn find_quick_add_spans(input: &str, now: LocalDateTime) -> Vec<QuickAddSpan> {
    let links = raw_note_links(input);
    let mut spans = link_spans(&links);

    if let Some(phrase) =
        raw_date_phrase(input, now).filter(|p| !is_inside_link(&links, p.range.start))
    {
        spans.push(RawSpan {
            range: phrase.range,
            kind: SpanKind::DatePhrase,
        });
    }
    if let Some((range, _)) = find_repeat_phrase_range(input, now.date(), now)
        .filter(|(range, _)| !is_inside_link(&links, range.start))
    {
        spans.push(RawSpan {
            range,
            kind: SpanKind::Repeat,
        });
    }

    let sigils = [
        (Marker::Priority, SpanKind::Priority),
        (Marker::Project, SpanKind::Project),
        (Marker::Tag, SpanKind::Tag),
    ];
    for (marker, kind) in sigils {
        for range in marker.runs(input) {
            if !is_inside_link(&links, range.start) {
                spans.push(RawSpan { range, kind });
            }
        }
    }

    drop_overlaps(spans)
        .into_iter()
        .map(|span| QuickAddSpan {
            start: utf16_index(input, span.range.start),
            end: utf16_index(input, span.range.end),
            kind: span.kind,
        })
        .collect()
}

/// Whether `input` holds any quick-add syntax, parseable marker or not.
pub fn has_special_syntax(input: &str, now: LocalDateTime) -> bool {
    [Marker::Priority, Marker::Project, Marker::Tag]
        .into_iter()
        .any(|marker| !marker.runs(input).is_empty())
        || !raw_note_links(input).is_empty()
        || raw_date_phrase(input, now).is_some()
        || find_repeat_phrase_range(input, now.date(), now).is_some()
}

/// The canonical cadence a half-typed "every ..." completes to, or `None`.
/// Case-insensitive: the cadence starts with the lowercased `query`.
pub fn predict_repeat_completion(query: &str) -> Option<String> {
    let lower = query.to_lowercase();
    REPEAT_PHRASES
        .into_iter()
        .find(|phrase| phrase.starts_with(&lower))
        .map(str::to_owned)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn marker_runs_follow_match_all() {
        assert_eq!(
            Marker::Priority.runs("task !nope !low"),
            vec![5..10, 11..15]
        );
        assert_eq!(Marker::Priority.runs("!a!b !c"), vec![0..2, 5..7]);
        assert_eq!(Marker::Tag.runs(" #a/b/ #c/-"), vec![1..5, 7..9]);
        assert!(Marker::Project.runs("1+2 is C++").is_empty());
    }

    #[test]
    fn note_links_need_both_brackets() {
        let links = raw_note_links("[[[a]] [[ ]] [[b");
        let ranges: Vec<_> = links.iter().map(|link| link.range.clone()).collect();
        assert_eq!(ranges, vec![1..6, 7..12]);
        assert_eq!(links[1].title, "");
    }
}
